sdks/python/apache_beam/utils/transform_service_launcher.py

```diff
@@ -0,0 +1,257 @@
+          'since could not find the Google Cloud application default '
+          'credentials file at the expected location %s.' %
+          application_default_path_file)
+
+    self._environmental_variables = {}
+    self._environmental_variables['CREDENTIALS_VOLUME'] = credentials_dir
+    self._environmental_variables['TRANSFORM_SERVICE_PORT'] = str(port)
+    self._environmental_variables['BEAM_VERSION'] = beam_version
+
+    self._docker_compose_start_command_prefix = []
+    self._docker_compose_start_command_prefix.append('docker-compose')
+    self._docker_compose_start_command_prefix.append('-p')
+    self._docker_compose_start_command_prefix.append(project_name)
+    self._docker_compose_start_command_prefix.append('-f')
+    self._docker_compose_start_command_prefix.append(compose_file)
+
+  def _get_channel(self):
+    channel_options = [("grpc.max_receive_message_length", -1),
+                       ("grpc.max_send_message_length", -1)]
+    if hasattr(grpc, 'local_channel_credentials'):
+      # TODO: update this to support secure non-local channels.
+      return grpc.secure_channel(
+          self._address,
+          grpc.local_channel_credentials(),
+          options=channel_options)
+    else:
+      return grpc.insecure_channel(self._address, options=channel_options)
+
+  def __enter__(self):
+    self.start()
+    self.wait_till_up(-1)
+
+    self._channel = self._get_channel()
+
+    from apache_beam import external
+    # with channel_factory_fn() as channel:
+    return external.ExpansionAndArtifactRetrievalStub(self._channel.__enter__())
+
+  def __exit__(self, *args):
+    self.shutdown()
+    self._channel.__exit__(*args)
+
+  def _run_docker_compose_command(self, command, output_override=None):
+    cmd = []
+    cmd.extend(self._docker_compose_start_command_prefix)
+    cmd.extend(command)
+
+    myenv = os.environ.copy()
+    myenv.update(self._environmental_variables)
+
+    process = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, env=myenv)
+    std_out, _ = process.communicate()
+
+    if output_override:
+      output_override.write(std_out)
+    else:
+      print(std_out.decode(errors='backslashreplace'))
+
+  def start(self):
+    with self._launcher_lock:
+      self._run_docker_compose_command(['up', '-d'])
+
+  def shutdown(self):
+    with self._launcher_lock:
+      self._run_docker_compose_command(['down'])
+
+  def status(self):
+    with self._launcher_lock:
+      self._run_docker_compose_command(['ps'])
+
+  def wait_till_up(self, timeout_ms):
+    channel = self._get_channel()
+
+    # Waiting till the service is up.
+    channel_ready = grpc.channel_ready_future(channel)
+    wait_secs = .1
+    start_time = time.time()
+    while True:
+      if (time.time() - start_time) * 1000 > timeout_ms > 0:
+        raise ValueError(
+            'Transform service did not start in %s seconds.' %
+            (timeout_ms / 1000))
```

Comment:
i think this timeout exception is never reached if we pass `timeout_ms=-1` like we do in main: `launcher.wait_till_up(-1)` (can lead to infinite loop).

use `_DEFAULT_START_WAIT_TIMEOUT` here if the timeout is <0?